testsetups/ocr.go

```diff
@@ -0,0 +1,210 @@
+		Str("Round Timeout", t.Inputs.RoundTimeout.String()).
+		Int("Number of OCR Contracts", len(t.ocrInstances)).
+		Msg("Starting OCR Soak Test")
+
+	testContext, testCancel := context.WithTimeout(context.Background(), t.Inputs.TestDuration)
+	defer testCancel()
+
+	// Test Loop
+	roundNumber := 1
+	for {
+		select {
+		case <-testContext.Done():
+			log.Info().Msg("Soak Test Complete")
+			return
+		default:
+			log.Info().Int("Round Number", roundNumber).Msg("Starting new Round")
+			adapterValue := t.changeAdapterValue(roundNumber)
+			t.waitForRoundToComplete(roundNumber)
+			t.checkLatestRound(adapterValue, roundNumber)
+			roundNumber++
+		}
+	}
+}
+
+// ensureValues ensures that all values needed to run the test are present
+func (t *OCRSoakTest) ensureInputValues() {
+	inputs := t.Inputs
+	Expect(inputs.NumberOfContracts).Should(BeNumerically(">=", 1), "Expecting at least 1 OCR contract")
+	Expect(inputs.ChainlinkNodeFunding.Float64()).Should(BeNumerically(">", 0), "Expecting non-zero chainlink node funding amount")
+	Expect(inputs.TestDuration).Should(BeNumerically(">=", time.Minute*1), "Expected test duration to be more than a minute")
+	Expect(inputs.RoundTimeout).Should(BeNumerically(">=", time.Second*15), "Expected test duration to be more than 15 seconds")
+}
+
+// changes the mock adapter value for OCR instances to retrieve answers from
+func (t *OCRSoakTest) changeAdapterValue(roundNumber int) int {
+	adapterValue := 0
+	if roundNumber%2 == 1 {
+		adapterValue = t.Inputs.StartingAdapterValue
+	} else {
+		adapterValue = t.Inputs.StartingAdapterValue * 25
+	}
+	By("Setting adapter responses",
+		actions.SetAllAdapterResponsesToTheSameValue(adapterValue, t.ocrInstances, t.chainlinkNodes, t.mockServer))
+	log.Debug().
+		Int("New Value", adapterValue).
+		Int("Round Number", roundNumber).
+		Msg("Changed Mock Server Adapter Value for New Round")
+	return adapterValue
+}
+
+// waits for the specified round number to complete on all deployed OCR instances
+func (t *OCRSoakTest) waitForRoundToComplete(roundNumber int) {
+	for _, ocrInstance := range t.ocrInstances {
+		report := t.TestReporter.Reports[ocrInstance.Address()]
+		ocrRound := contracts.NewOffchainAggregatorRoundConfirmer(
+			ocrInstance,
+			big.NewInt(int64(roundNumber)),
+			t.Inputs.RoundTimeout,
+			report,
+		)
+		t.defaultNetwork.AddHeaderEventSubscription(ocrInstance.Address(), ocrRound)
+	}
+	err := t.defaultNetwork.WaitForEvents()
+	Expect(err).ShouldNot(HaveOccurred(), "Error while waiting for OCR round number %d to complete", roundNumber)
+}
+
+// checks on all OCR instances that they all received the correct answer from the latest round
+func (t *OCRSoakTest) checkLatestRound(expectedValue, roundNumber int) {
+	var roundAnswerGroup sync.WaitGroup
+	roundAnswerChannel := make(chan latestRoundAnswer, len(t.ocrInstances))
+	for _, ocrInstance := range t.ocrInstances {
+		roundAnswerGroup.Add(1)
+		ocrInstance := ocrInstance
+		go func() {
```

Comment:
here maybe you could do something like
```
    done := make(chan interface{})
    go func() {
      defer GinkgoRecover()

      content := book.RawContent()
      Expect(book.Pages).To(Equal(2783))

      //this might be quite expensive and will block...
      err := book.SetFontSize(28)
      Expect(err).NotTo(HaveOccurred())

      Expect(book.Pages).To(BeNumerically(">", 2783))
      Expect(book.RawContent()).To(Equal(content))

      close(done)
    }()

    Eventually(done).Should(BeClosed())
```

from https://onsi.github.io/ginkgo/#patterns-for-asynchronous-testing

Reply:
The other defer is to control the wait group - assuming there is no failure - which I just pushed a bit of a cleaner version.